src/main/scala/karazin/scala/users/group/week2/homework/Homework.scala

```diff
@@ -34,28 +34,48 @@ object Homework:
       !(this < that)
 
     @targetName("addition")
-    infix def +(that: Rational): Rational = ???
+    infix def +(that: Rational): Rational =
+      new Rational(
+        this.numer * that.denom + this.denom * that.numer,
+        this.denom * that.denom
+      )
 
     @targetName("negation")
-    infix def unary_- : Rational = ???
+    infix def unary_- : Rational =
+      new Rational(-this.numer, this.denom)
 
     @targetName("substraction")
-    infix def -(that: Rational): Rational = ???
+    infix def -(that: Rational): Rational =
+      this + -that
 
     @targetName("multiplication")
-    infix def *(that: Rational): Rational = ???
+    infix def *(that: Rational): Rational =
+      new Rational(this.numer * that.numer, this.denom * that.denom)
 
     @targetName("devision")
-    infix def /(that: Rational): Rational = ???
+    infix def /(that: Rational): Rational =
+      require(that.numer != 0, "Zero Division Error")
+      val denom = this.denom * that.numer
+
+      if denom < 0
+      then new Rational(-(this.numer * that.denom), abs(denom))
+      else new Rational(this.numer * that.denom, this.denom * that.numer)
 
     override def toString: String = s"${this.numer}/${this.denom}"
 
+    def toDouble: Double = this.numer.toDouble / this.denom.toDouble
+
     private def gcd(a: Int, b: Int): Int =
       if b == 0 then a else gcd(b, a % b)
 
     private lazy val g = gcd(abs(x), y)
 
-    override def equals(other: Any): Boolean = ???
+    override def equals(other: Any): Boolean = other match {

```

Comment:
Every time you implement equals you have to implement hashCode.
https://www.baeldung.com/java-equals-hashcode-contracts
https://alvinalexander.com/scala/how-to-define-equals-hashcode-methods-in-scala-object-equality/